### PR TITLE
check `FnDef` return type for WF

### LINKED ISSUE #84533: Functions, closures, and HRTB-trait-objects can implement traits such that validity of associated types is never checked.
_Make sure to also read through the next few comments of mine, where I present some other, significantly different examples of related problems (that’s where HRTBs become relevant, and closures, fn-pointers and trait objects play a role). I also explain the T-lang tag there with a question on changing the meaning of HRTBs that include associated types. Possibly it’s even a sensible thing to split this issue up into two issues._

So if I write a function such as
```rust
fn foo<'b, 'a>() -> PhantomData<&'b &'a ()> {
    PhantomData
}
```
then `foo`’s return type is only valid if `'a: 'b`. When we call the function this is enforced a the call site, e.g.
```rust
fn caller<'b, 'a>() {
    foo::<'b, 'a>();
}
```
doesn’t work, with
```rust
error[E0491]: in type `&'b &'a ()`, reference has a longer lifetime than the data it references
 --> src/lib.rs:8:5
  |
8 |     foo::<'b, 'a>();
  |     ^^^^^^^^^^^^^^^
  |
note: the pointer is valid for the lifetime `'b` as defined on the function body at 7:11
 --> src/lib.rs:7:11
  |
7 | fn caller<'b, 'a>() {
  |           ^^
note: but the referenced data is only valid for the lifetime `'a` as defined on the function body at 7:15
 --> src/lib.rs:7:15
  |
7 | fn caller<'b, 'a>() {
  |               ^^

error: aborting due to previous error
```
([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=0735b1090d7b275ab12a940867788c5f))

However, if we _just_ instantiate the lifetime parameters, there’s apparently no check being performed at all, i.e.
```rust
fn caller<'b, 'a>() {
    foo::<'b, 'a>;
}
```
compiles just fine. ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=efe80b69733df331716f18eea6f4130c)).

This seems questionable, the type of `foo::<'b, 'a>` implements `FnOnce` with `Output = PhantomData<&'b &'a ()>`, and it is in fact possible to use this output type to circumvent the borrow checker:
```rust
use std::marker::PhantomData;

fn foo<'b, 'a>() -> PhantomData<&'b &'a ()> {
    PhantomData
}

fn extend_lifetime<'a, 'b, T: ?Sized>(x: &'a T) -> &'b T {
    let f = foo::<'b, 'a>;
    f.baz(x)
}

trait Foo<'a, 'b, T: ?Sized> {
    fn baz(self, s: &'a T) -> &'b T;
}
impl<'a, 'b, R, F, T: ?Sized> Foo<'a, 'b, T> for F
where
    F: Fn() -> R,
    R: ProofForConversion<'a, 'b, T>,
{
    fn baz(self, s: &'a T) -> &'b T {
        self().convert(s)
    }
}

trait ProofForConversion<'a, 'b, T: ?Sized> {
    fn convert(self, s: &'a T) -> &'b T;
}
impl<'a, 'b, T: ?Sized> ProofForConversion<'a, 'b, T> for PhantomData<&'b &'a ()> {
    fn convert(self, s: &'a T) -> &'b T {
        s
    }
}

fn main() {
    let d;
    {
        let x = "Hello World".to_string();
        d = extend_lifetime(&x);
    }
    println!("{}", d);
}
```
```
�ZV&V
```
([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=5167c528f2e0dfa4a4fbe5af4e20c081))

@rustbot modify labels: A-lifetimes, A-typesystem, T-compiler
and someone please add “I-unsound 💥”.

### PR DESCRIPTION
**Pull request by lcnr**
better version of #106807, fixes #84533 (mostly). It's not perfect given that we still ignore WF requirements involving bound regions but I wasn't able to quickly write an example, so even if theoretically exploitable, it should be far harder to trigger.

This is strictly more restrictive than checking the return type for WF as part of the builtin `FnDef: FnOnce` impl (#106807) and moving to this approach in the future will not break any code.

~~It also agrees with my theoretical view of how this should behave~~

r? types

### CODE REVIEW

**Comment by lcnr** [comment]
@craterbot check

**Comment by lcnr** [comment]
didn't push a change which makes this check more powerful

@craterbot cancel

@bors try

**Comment by lcnr** [comment]
@craterbot abort

**Comment by lcnr** [comment]
@craterbot check

**Comment by lcnr** [comment]
@craterbot check end=8ce16caa975da810c768ab50cc10f47a50531548 crates=https://crater-reports.s3.amazonaws.com/pr-115538/retry-regressed-list.txt

**Comment by lqd** [comment]
fingers crossed

@craterbot mode=check-only start=master#8cfaf70c320297dd1831a12351fcfeab7a231328 end=try#8ce16caa975da810c768ab50cc10f47a50531548 crates=https://crater-reports.s3.amazonaws.com/pr-115538/retry-regressed-list.txt

**Comment by lqd** [comment]
fingers crossed²

@craterbot name=pr-115538-2 mode=check-only start=master#8cfaf70c320297dd1831a12351fcfeab7a231328 end=try#8ce16caa975da810c768ab50cc10f47a50531548 crates=https://crater-reports.s3.amazonaws.com/pr-115538/retry-regressed-list.txt

**Comment by lqd** [comment]
fingers crossed^3

@craterbot run name=pr-115538-3 mode=check-only start=master#8cfaf70c320297dd1831a12351fcfeab7a231328 end=try#8ce16caa975da810c768ab50cc10f47a50531548 crates=https://crater-reports.s3.amazonaws.com/pr-115538/retry-regressed-list.txt

**Comment by jackh726** [comment]
I'm tempted to just r+ this...crater shows no regressions. We should *probably* do an FCP though.

**Comment by jackh726** [comment]
@rfcbot merge

Should we check `FnDef` return types for WF? This takes a step forward to mostly #84533. Crater shows no regressions.

**Comment by aliemjay** [comment]
Note that this can break valid code. The only way I can think of is when the fn item is used in an RPIT in the following example. Also note the new inconsistency between closures and fn items.
```rust
fn id<X>(x: X) -> X { x }

pub fn get<T>() -> impl Fn(&'static T) -> &'static T {
    id //~ OK -> ERROR; requires `T: 'static`
    //|x| id(x) // still OK
}
```

I think this breakage is not necessary for soundness (see below for an alternative approach) but given that it is arguably a niche case, and the crater results are clear, I'm fine with it.
@rfcbot reviewed

In my opinion, a better way to deal with the unsoundness is to lint the erroneous functions at the definition site:
```rust
fn foo<'c, 'd>() -> PhantomData<&'c &'d ()> { //~ ERROR; requires '`d: 'c`
    PhantomData
}
```
I think this is more correct because it deals with late bound lifetimes as well, forcing them to be downgraded to be early-bound because of the required where clause. It is also consistent with how we check trait impls for well-formedness - in this example we error at the impl site, not where the impl is used.
```rust
impl<'c, 'd> FnOnce<()> for typeof(foo::<'c, 'd>) {
    type Output = PhantomData<&'c &'d ()>; //~ ERROR; requires `'d: 'c`
    // ...
}
```
Additionally, I believe it would be easy to implement as lint in wfcheck: just do wfcheck of the function without the implied bounds of the return type, if failed, redo the check now with full implied bounds, compare the results and lint in case of disparity.

**Comment by lcnr** [comment]
thought about this a week ago or so but never got to actually finish that:

```rust
fn foo<'b, 'a>() -> PhantomData<&'b &'a ()> {
    PhantomData
}

fn bar<'b, 'a>(x: &'a ()) -> PhantomData<&'b &'a ()> {
    PhantomData
}
```
from my pov this gets lowered to the following types and impls:
```rust
struct foo<'b, 'a>;
impl<'b, 'a> FnOnce<()> for foo<'b, 'a>
where
    #[implied] WF(fn() -> PhantomData<&'b &'a ()>),    
{
    type Assoc = &'b &'a ();
}

struct bar<'b>;
impl<'b, 'a> FnOnce<(&'a (),)> for bar<'b>
where
    #[implied] WF(fn(&'a ()) -> PhantomData<&'b &'a ()>),
```
So this PR does not agree with my ideal view on how this should be handled. Afaict the proposal by @aliemjay is "stop considering implied bounds from the return type" and is not necessarily where we want to end up. If possible without causing too much breakage it would give us more space in the future when making implied bounds explicit. I think it will be fine even with implied bounds from the return type but I didn't think about this in-depth.

**Comment by BoxyUwU** [comment]
Happy that this will be forwards compatible with checking stuff as part of the impl not wfness of the fndef since that seems much more reasonable to me. If it wasn't I would want to hold off on this.

I think aliemjay's proposal of not introducing implied bounds from wfness of ret ty would probably be a simple to verify solution to this unsoundness, have we done a crater run of it? It's trivial to come up with artificial examples where that fix would break code that _could_ reasonably work, and i'd personally be inclined to say that it ought to. If it doesn't have too many regressions though imo that solution and then one day re-adding implied bounds there and checking everything in the builtin impl would be best since it removes any unsoundness short term.

**Comment by BoxyUwU** [comment]
@rfcbot concern as an alternate short term solution how breaking would removing implied bounds derived from return types be?

**Comment by lcnr** [comment]
removing implied bounds for return types is incredibly breaking: https://github.com/lcnr/rust/commit/bf3580373e03f35a9b23bffe67729adf4e61b6d3

bootstrapping results in 67 errors, here some examples:
```rust
pub const unsafe fn from_raw_parts<'a, T>(data: *const T, len: usize) -> &'a [T] { .. }

impl<T: ?Sized> RangeBounds<T> for RangeFull {
    fn start_bound(&self) -> Bound<&T> { .. }
}

pub trait AsRef<T: ?Sized> {
    fn as_ref(&self) -> &T;
}


pub trait Borrow<Borrowed: ?Sized> {
    fn borrow(&self) -> &Borrowed;
}
```

**Comment by BoxyUwU** [comment]
@rfcbot resolve as an alternate short term solution how breaking would removing implied bounds derived from return types be?

File: compiler/rustc_trait_selection/src/traits/wf.rs [hunk 1]
```diff
@@ -767,12 +778,12 @@ impl<'a, 'tcx> WfPredicates<'a, 'tcx> {
...
                     // checking those
 
-                    let defer_to_coercion = self.tcx().features().object_safe_for_dispatch;
+                    let defer_to_coercion = tcx.features().object_safe_for_dispatch;
 
                     if !defer_to_coercion {
                         let cause = self.cause(traits::WellFormed(None));
                         let component_traits = data.auto_traits().chain(data.principal_def_id());
-                        let tcx = self.tcx();
+                        let tcx = tcx;
```
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/traits/wf.rs:786]:
```suggestion
```

**Review by compiler-errors — APPROVED** [review]
r=me after nit

**Comment by Dylan-DPC** [comment]
@lcnr any updates on this? this just requires a conflict resolution and a tiny nit fix

File: compiler/rustc_trait_selection/src/traits/wf.rs [hunk 2]
```diff
@@ -707,6 +709,16 @@ impl<'a, 'tcx> TypeVisitor<TyCtxt<'tcx>> for WfPredicates<'a, 'tcx> {
             }
 
             ty::FnDef(did, args) => {
+                // HACK: Check the return type of function definitions for
+                // well-formedness to mostly fix #84533. This is still not
+                // perfect and there may be ways to abuse the fact that we
+                // ignore requirements with escaping bound vars. That's a
+                // more general issue however.
+                //
+                // FIXME(eddyb) add the type to `walker` instead of recursing.
```
Comment (lcnr) [inline compiler/rustc_trait_selection/src/traits/wf.rs:718]: this fixme is outdated.

**Pull request merged**